File: go/bowling.go
package bowling

import (
	"errors"
)

var (
	ErrRollNegative   = errors.New("Negative roll is invalid")
	ErrRollExceedPins = errors.New("Pin count exceeds pins on the lane")
	ErrRollGameOver   = errors.New("Cannot roll after game is over")
	ErrCannotScore    = errors.New("cannot scored")
)

// Define the Game type here.
type Game struct {
	rolls []int
}

func NewGame() *Game {
	return &Game{}
}

func (g *Game) Roll(pins int) error {
	if pins < 0 {
		return ErrRollNegative
	}

	if pins > 10 {
		return ErrRollExceedPins
	}

	if len(g.rolls) == 20 {
		if g.rolls[18]+g.rolls[19] != 10 {
			return ErrRollGameOver
		}
	}

	if len(g.rolls) == 21 {
		if g.rolls[18] != 10 {
			return ErrRollGameOver
		}
		if g.rolls[20] != 10 && g.rolls[20]+pins > 10 {
			return ErrRollExceedPins
		}
	}

	if len(g.rolls) == 22 {
		return ErrRollGameOver
	}

	if len(g.rolls) < 20 && len(g.rolls)%2 == 1 && pins+g.rolls[len(g.rolls)-1] > 10 {
		return ErrRollExceedPins
	}

	if pins == 10 {
		g.rolls = append(g.rolls, 10)
		if len(g.rolls) < 20 {
			g.rolls = append(g.rolls, 0)
		}
	} else {
		g.rolls = append(g.rolls, pins)
	}
	return nil
}

func (g *Game) Score() (int, error) {
	if len(g.rolls) < 20 || g.rolls[18] == 10 && len(g.rolls) != 22 || g.rolls[18] != 10 && g.rolls[18]+g.rolls[19] == 10 && len(g.rolls) != 21 {
		return 0, ErrCannotScore
	}

	s := 0
	for i := 0; i < 20; i += 2 {
		if g.rolls[i] == 10 {
			s += 10
			if g.rolls[i+2] == 10 {
				if i == 18 {
					s += g.rolls[i+2] + g.rolls[i+3]
				} else {
					s += 10 + g.rolls[i+4]
				}
			} else {
				s += g.rolls[i+2] + g.rolls[i+3]
			}
		} else if g.rolls[i]+g.rolls[i+1] == 10 {
			s += 10 + g.rolls[i+2]
		} else {
			s += g.rolls[i] + g.rolls[i+1]
		}
	}

	return s, nil
}

File: ocaml/bowling.ml
open Base

type t = { rolls : int list; frame : int }

let new_game = { rolls = []; frame = 0 }

let last_frame = function
  | [] -> (0, 0)
  | [ a ] -> (a, 0)
  | a :: b :: _ -> (a, b)

let roll r { rolls; frame } =
  match (r, frame, last_frame rolls) with
  | r, _, _ when r < 0 -> Error "Negative roll is invalid"
  | r, _, _ when r > 10 -> Error "Pin count exceeds pins on the lane"
  | r, f, (a, _) when f < 20 && f % 2 = 1 && r + a > 10 ->
      Error "Pin count exceeds pins on the lane"
  | _, 20, (a, b) when a + b < 10 -> Error "Cannot roll after game is over"
  | _, 21, (a, b) when b <> 10 -> Error "Cannot roll after game is over"
  | 10, 21, (a, _) when a <> 10 -> Error "Pin count exceeds pins on the lane"
  | r, 21, (a, 10) when a <> 10 && r + a > 10 ->
      Error "Pin count exceeds pins on the lane"
  | _, 22, _ -> Error "Cannot roll after game is over"
  | 10, f, _ when f < 20 && f % 2 = 0 ->
      Ok { rolls = 10 :: rolls; frame = frame + 2 }
  | _ -> Ok { rolls = r :: rolls; frame = frame + 1 }

let score g =
  let rec score' acc f rolls =
    if f < 10 then
      match rolls with
      | 10 :: (a :: b :: _ as rs) -> score' (acc + 10 + a + b) (f + 1) rs
      | a :: b :: (c :: _ as rs) when a + b = 10 ->
          score' (acc + 10 + c) (f + 1) rs
      | a :: b :: rs when a + b < 10 -> score' (acc + a + b) (f + 1) rs
      | _ -> Error "Score cannot be taken until the end of the game"
    else Ok acc
  in
  List.rev g.rolls |> score' 0 0
